dev-tools/test.sh

```diff
@@ -16,23 +16,61 @@ require_database
 # Set dummy key to enable SUMM.AI during testing
 export INTEGREAT_CMS_SUMM_AI_API_KEY="dummy"
 
-# If only particular tests should be run, test path can be passed as CLI argument
-TEST_PATH=$1;
-
-if [[ -z "$TEST_PATH" ]]; then
-    echo -e "Running all tests..." | print_info
-    deescalate_privileges pipenv run pytest --disable-warnings --quiet --numprocesses=auto --cov=integreat_cms --cov-report html
-elif [[ -e "$TEST_PATH" ]]; then
-    echo -e "Running tests in ${TEST_PATH}..." | print_info
-    deescalate_privileges pipenv run pytest "$TEST_PATH" --quiet --disable-warnings --numprocesses=auto
+# Parse given command line arguments
+while [[ $# -gt 0 ]]; do
+    case $1 in
+        # If only tests affected by recent changed should be run, --changed can be passed as a flag
+        --changed) CHANGED=1;shift;;
+        # If only particular tests should be run, test path can be passed as CLI argument
+        *) TEST_PATH=$1;shift;;
+    esac
+done
+
+# The default pytests args we use
+PYTEST_ARGS=("--quiet" "--disable-warnings" "--numprocesses=auto")
+
+# Check if --changed flag was passed
+if [[ -n "${CHANGED}" ]]; then
+    # Check if .testmondata file exists
+    if [[ -f ".testmondata" ]]; then
+        # Only run changed tests and don't update dependency database
+        PYTEST_ARGS+=("--testmon-nocollect")
+        CHANGED_MESSAGE=" affected by recent changes"
+    else
+        # Inform that all tests will be run
+        echo -e "\nIt looks like you have not run pytest without the \"--changed\" flag before." | print_warning
+        echo -e "Pytest has to build a dependency database by running all tests without the flag once.\n" | print_warning
+        # Override test path argument
+        unset TEST_PATH
+        # Tell testmon to run all tests and collect data
+        PYTEST_ARGS+=("--testmon-noselect")
+    fi
 else
+    # Run all tests, but update list of tests
+    PYTEST_ARGS+=("--testmon-noselect")
+fi
+
+# Determine whether coverage data should be collected
+if [[ -z "${CHANGED}" && -z "${TEST_PATH}" ]]; then
+    PYTEST_ARGS+=("--cov=integreat_cms" "--cov-report html")
+fi
+
+# Check whether test path exists
+if [[ -e "${TEST_PATH}" ]]; then
+    # Adapt message and append to pytest arguments
+    TEST_MESSAGE=" in ${TEST_PATH}"
+    PYTEST_ARGS+=("${TEST_PATH}")
+elif [[ -n "${TEST_PATH}" ]]; then
+    # If the test path does not exist but was non-zero, show an error
     echo -e "${TEST_PATH}: No such file or directory" | print_error
     exit 1
 fi
 
+echo -e "Running all tests${TEST_MESSAGE}${CHANGED_MESSAGE}..." | print_info
+eval "deescalate_privileges pipenv run pytest ${PYTEST_ARGS[*]}"
```

Comment:
Hmm, I'm not convinced that this is a valid use case for `eval`...
Maybe we could debug the problems with the previous solution?
Since `#!/bin/bash` is used as shebang, your default shell should have no influence on the fact that the script is executed by bash... but it could be something platform-specific, what's the output of your `/bin/bash --version`?
Could you check out my commit and insert `set -x` somewhere above the pytest command? This should print out what exactly is being run.
E.g. when I run ` ./dev-tools/test.sh --changed tests/api`, it prints out the following:
```
Running all tests in tests/api affected by recent changes...
+ IFS=
+ read -r line
+ deescalate_privileges pipenv run pytest --quiet --disable-warnings --numprocesses=auto --testmon-nocollect tests/api
++ id -u
+ [[ 1000 == 0 ]]
+ env pipenv run pytest --quiet --disable-warnings --numprocesses=auto --testmon-nocollect tests/api
```

Reply:
Ah, good call - the issue was the argument `--cov-report html`, which got auto-escaped. Replaces it with `--cov-report=html`, now everything works as intended.

Thanks btw, didn't know about `set -x`

Reply:
Ahh ok makes sense, thanks for figuring that out! :+1: